oxide/lib.go

```diff
@@ -22,75 +22,108 @@ const TokenEnvVar = "OXIDE_TOKEN"
+	// Base URL of the Oxide API including the scheme. For example,
+	// https://api.oxide.computer.
+	Host string
+
+	// Oxide API authentication token.
+	Token string
+
+	// A custom HTTP client to use for the Client instead of the default.
+	HTTPClient *http.Client
+
+	// A custom user agent string to add to every request instead of the
+	// default.
+	UserAgent string
+}
+
 // Client which conforms to the OpenAPI3 specification for this service.
 type Client struct {
-	// The endpoint of the server conforming to this interface, with scheme,
-	// https://api.oxide.computer for example.
+	// Base URL of the Oxide API including the scheme. For example,
+	// https://api.oxide.computer.
 	server string
 
-	// Client is the *http.Client for performing requests.
+	// Oxide API authentication token.
+	token string
+
+	// HTTP client to make API requests.
 	client *http.Client
 
-	// token is the API token used for authentication.
-	token string
+	// The user agent string to add to every API request.
+	userAgent string
 }
 
-// NewClient creates a new client for the Oxide API.
-// You need to pass in your API token to create the client.
-func NewClient(token, userAgent, host string) (*Client, error) {
-	if token == "" {
-		return nil, fmt.Errorf("you need to pass in an API token to create the client")
+// NewClient creates a new client for the Oxide API. To authenticate with
+// environment variables, set OXIDE_HOST and OXIDE_TOKEN accordingly. Pass in a
+// non-nil *Config to set the various configuration options on a Client. When
+// setting the host and token through the *Config, these will override any set
+// environment variables.
+func NewClient(cfg *Config) (*Client, error) {
+	token := os.Getenv(TokenEnvVar)
+	server := os.Getenv(HostEnvVar)
+	userAgent := defaultUserAgent()
+	httpClient := &http.Client{
+		Timeout: 600 * time.Second,
 	}
 
-	baseURL, err := parseBaseURL(host)
-	if err != nil {
-		return nil, err
+	// Layer in the user-provided configuration if provided.
+	if cfg != nil {
+		if cfg.Host != "" {
+			server = cfg.Host
+		}
+
+		if cfg.Token != "" {
+			token = cfg.Token
+		}
+
+		if cfg.UserAgent != "" {
+			userAgent = cfg.UserAgent
+		}
+
+		if cfg.HTTPClient != nil {
+			httpClient = cfg.HTTPClient
+		}
 	}
 
-	client := &Client{
-		server: baseURL,
-		token:  token,
+	var errServer error
+	server, err := parseBaseURL(server)
+	if err != nil {
+		errServer = fmt.Errorf("failed parsing host address: %w", err)
 	}
 
-	// Ensure the server URL always has a trailing slash.
-	if !strings.HasSuffix(client.server, "/") {
-		client.server += "/"
+	var errToken error
+	if token == "" {
+		errToken = errors.New("token is required")
 	}
 
-	uat := userAgentTransport{
-		base:      http.DefaultTransport,
-		userAgent: userAgent,
-		client:    client,
+	// To aggregate the validation errors above.
+	if err := errors.Join(errServer, errToken); err != nil {
+		return nil, fmt.Errorf("invalid client configuration: %w", err)
```

Comment:
Perhaps the error may look nicer if we add a newline here :)

Reply:
Done!